### PR TITLE
Have ability to search made request by condition

### PR DESCRIPTION
**Pull request by drudoi**
Issue link: https://github.com/mcustiel/phiremock/issues/23

Tests result:
```
$ vendor/bin/codecept -c tests run
Codeception PHP Testing Framework v2.3.5
Powered by PHPUnit 6.2.4 by Sebastian Bergmann and contributors.
Running exec php /opt/phiremock/tests/tests/../../bin/phiremock --port 8086

Acceptance Tests (69) -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
✔ BodyConditionCest: Create an expectation that checks body using isequalto (0.24s)
✔ BodyConditionCest: Create an expectation that checks body using matches (0.02s)
✔ BodyConditionCest: See if request fails when an invalid matcher is specified (0.02s)
✔ BodyConditionCest: Check if the request fails when and invalid value is specified (0.01s)
✔ BodyConditionCest: See if mocking based in request body pattern works (0.01s)
✔ BodyConditionCest: See if mocking based in request body equality works (0.01s)
✔ BodyConditionCest: See if mocking based in request body case insensitive equality works (0.01s)
✔ BodyJsonCest: Create an expectation with a json body defined as array (0.01s)
✔ BodyJsonCest: Create an expectation with a json body defined as object (0.01s)
✔ BodySpecificationCest: Create an expectation with a valid body (0.01s)
✔ BodySpecificationCest: Create an expectation with an empty body (0.01s)
✔ ClientCest: Should create an expectation test (0.02s)
✔ ClientCest: Should create an expectation and receive it in the list (0.02s)
✔ ClientCest: Should list several expectations (0.02s)
✔ ClientCest: Should create an expectation test with fluent interface (3.52s)
✔ ClientCest: Count executions test (0.03s)
✔ ClientCest: List executions test (0.03s)
✔ ClientCest: Count executions when no expectation is set (0.06s)
✔ ClientCest: Contains matcher should work (2.51s)
✔ ClientCest: Full url should be evaluated (2.52s)
✔ DelaySpecificationCest: Create an expectation with a valid delay specification (0.02s)
✔ DelaySpecificationCest: Create an expectation with a negative delay specification (0.04s)
✔ DelaySpecificationCest: Create an expectation with an invalid delay specification (0.01s)
✔ DelaySpecificationCest: Mock a request with delay (2.01s)
✔ ExpectationCreationCest: Create an expectation that only checks url (0.02s)
✔ ExpectationCreationCest: Create an expectation that only checks method (0.02s)
✔ ExpectationCreationCest: Create an expectation that only checks body (0.01s)
✔ ExpectationCreationCest: Create an expectation that only checks headers (0.01s)
✔ ExpectationCreationCest: See if creation fails when request is empty (0.01s)
✔ ExpectationCreationCest: When response is empty in request, default should be used (0.01s)
✔ ExpectationCreationCest: See if creation fails when anything sent as request (0.01s)
✔ ExpectationCreationCest: See if creation fails when anything sent as response (0.01s)
✔ ExpectationCreationCest: Create an expectation with all possible option filled (0.01s)
✔ ExpectationListCest: Return empty list test (0.00s)
✔ ExpectationListCest: Return created expectation test (0.01s)
✔ HeadersConditionsCest: Create an expectation that checks one header using isequalto (0.01s)
✔ HeadersConditionsCest: Create an expectation that checks one header using matches (0.01s)
✔ HeadersConditionsCest: Fail when the matcher is invalid (0.01s)
✔ HeadersConditionsCest: Fail when the value is null (0.01s)
✔ HeadersConditionsCest: Create an expectation that checks more than one header (0.01s)
✔ HeadersConditionsCest: See if mocking based in one request header works (0.01s)
✔ HeadersConditionsCest: See if mocking based in several request headers works (0.01s)
✔ HeadersSpecificationCest: Create an specification with one header in response (0.01s)
✔ HeadersSpecificationCest: Create an specification with several headers in response (0.01s)
✔ HeadersSpecificationCest: Create a specification with no headers in response (0.01s)
✔ HeadersSpecificationCest: Fail when creating a specification with invalid headers (0.01s)
✔ MethodConditionCest: Create a specification with method post (0.01s)
✔ MethodConditionCest: Create a specification with method get (0.01s)
✔ MethodConditionCest: Create a specification with method put (0.01s)
✔ MethodConditionCest: Create a specification with method delete (0.01s)
✔ MethodConditionCest: Create a specification with method fetch (0.01s)
✔ MethodConditionCest: Create a specification with method options (0.01s)
✔ MethodConditionCest: Create a specification with method head (0.01s)
✔ MethodConditionCest: Create a specification with method patch (0.01s)
✔ MethodConditionCest: Create a specification that checks url using matches (0.01s)
✔ ProxyCest: Create an expectation with proxy to test (0.01s)
✔ ProxyCest: Proxy to given uri test (0.73s)
✔ ReplacementCest: Create an expectation with regex replacement from url (0.02s)
✔ ReplacementCest: Create an expectation with regex replacement from body (0.01s)
✔ ReplacementCest: Create an expectation with regex replacement from body and url (0.01s)
✔ ReplacementCest: Create an expectation with strict regex replacement from body and url (0.01s)
✔ ReplacementCest: Create an expectation without regex replacement (0.01s)
✔ StatusCodeSpecificationCest: Create a specification with a valid status code (0.01s)
✔ StatusCodeSpecificationCest: Create a specification with a default status code (0.01s)
✔ StatusCodeSpecificationCest: Fail when the status code is not set (0.01s)
✔ UrlConditionCest: Create an expectation that checks url using isequalto (0.01s)
✔ UrlConditionCest: Create an expectation that checks url using matches (0.01s)
✔ UrlConditionCest:  check if it fails when an invalid matcher is specified (0.01s)
✔ UrlConditionCest: Check if it fails when an invalid value is specified (0.01s)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Functional Tests (0) ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Unit Tests (0) ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 14.2 seconds, Memory: 16.00MB

OK (69 tests, 255 assertions)
```

### CODE REVIEW

File: src/Server/Config/router-config.php [hunk 1]
```diff
@@ -163,6 +163,28 @@
...
+                    ['goto' => 'verifyMethodIsPut'],
+                ],
+            ],
+        ],
+        'verifyMethodIsPut' => [
+            'condition' => [
+                'all-of' => [
+                    [
+                        'input-source' => ['method' => null],
+                        'matcher'      => ['isEqualTo' => 'PUT'],
```
Comment (drudoi) [inline src/Server/Config/router-config.php:175]: I've used PUT method for the callback, it isn't the best solution, but requests are processed by this (executions) section, and POST is already used. I didn't extended the URL as I see URLs are simple for all callbacks. 

POST could be used for this, but it needs URLs extension for both the added callback and count callback like `executions/list` and `executions/count`, or some other kind of separation (body param, headers, etc.).

File: src/Server/Actions/ListRequestsAction.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        $executions = [];
+        foreach ($this->requestsStorage->listRequests() as $request) {
+            if ($this->comparator->equals($request, $expectation)) {
+                $executions[] = [
+                    'method' => $request->getMethod(),
+                    'url' => (string) $request->getUri(),
+                    'headers' => $request->getHeaders(),
+                    'cookies' => $request->getCookieParams(),
+                    'body' => (string) $request->getBody(),
```
Comment (drudoi) [inline src/Server/Actions/ListRequestsAction.php:71]: I've used custom structure as I didn't find quick way to represent `ServerRequestInterface` object in JSON.

File: src/Server/Config/router-config.php (hunk 1 above)
Comment (mcustiel) [reply src/Server/Config/router-config.php:175]: Hey. I think having __phiremock/executions as a url could be a good idea. It represents what we are requesting and also we can use GET as verb, that also represents better the retrieval of the executions. What do you think?
If you agree I can do the change myself after merging.

File: src/Server/Actions/ListRequestsAction.php (hunk 2 above)
Comment (mcustiel) [reply src/Server/Actions/ListRequestsAction.php:71]: No problem. We can create a response value object for it. I can do it after merging.

File: src/Server/Config/router-config.php (hunk 1 above)
Comment (drudoi) [reply src/Server/Config/router-config.php:175]: Sorry, I didn't caught the idea. I've used `PUT __phiremock/executions` for the action. Do you propose use `GET __phiremock/executions` for this?
If it so, I don't think that GET method is better - see https://stackoverflow.com/a/983458.

Comment (mcustiel) [reply src/Server/Config/router-config.php:175]: I see. So it was a very good idea to have PUT as verb.

**Pull request merged**